typed/rosette/base-forms.rkt

```diff
@@ -0,0 +1,390 @@
+  (define (split-at* lst lengths)
+    (cond [(empty? lengths) (list lst)]
+          [else
+           (define-values [fst rst]
+             (split-at lst (first lengths)))
+           (cons fst (split-at* rst (rest lengths)))]))
+  )
+
+;; ----------------------------------------------------------------------------
+
+;; Expanding to Rosette Forms
+
+(begin-for-syntax
+  ;; TODO: fix this so it's not using hardcoded list of ids
+  (define (expand/ro e)
+    (define e+
+      (local-expand
+       e 'expression
+       (list #'ro:#%app #'ro:choose #'ro:synthesize #'ro:let #'ro:in-value
+             #'ro:assert #'ro:if #'ro:? #'ro:verify)))
+;    (displayln (stx->datum e+))
+    e+)
+  (define (mk-ro:-id id) (format-id id "ro:~a" id))
+  (current-host-lang mk-ro:-id))
+
+;; ----------------------------------------------------------------------------
+
+;; Declaring Types before Definitions
+
+(begin-for-syntax
+  (define type-decl-internal-id 'type-decl-internal-id)
+  (define type-decl-internal-id-for 'type-decl-internal-id-for)
+  (define-syntax-class id/type-decl
+    #:attributes [internal-id type]
+    [pattern x:id
+      ;; expand x in such a way that an unbound identifier
+      ;; won't be an error
+      #:with x* (local-expand #'x 'expression #false)
+      #:attr internal-id (syntax-property #'x* type-decl-internal-id)
+      #:when (attribute internal-id)
+      #:with type (typeof #'x*)]))
+
+(define-typed-syntax :
+  #:datum-literals [:]
+  [(_ x:id : τ) ≫
+   #:with x- (generate-temporary #'x)
+   --------
+   [≻ (define-syntax- x
+        (make-variable-like-transformer
+         (set-stx-prop/preserved
+          (set-stx-prop/preserved
+           (⊢ x- : τ)
+           type-decl-internal-id
+           (syntax-local-introduce #'x-))
+          type-decl-internal-id-for
+          (syntax-local-introduce #'x))))]])
+
+;; ----------------------------------------------------------------------------
+
+;; Define and Lambda
+
+(begin-for-syntax
+  (define-syntax-class ->
+    [pattern (~or (~datum →) (~datum ->))]))
+
+;; This new version of define handles keyword arguments,
+;; and also uses make-variable-like-transformer so that
+;; types are preserved accross modules.
+
+(define-typed-syntax define
+  #:datum-literals [:]
+  [(_ x:id : τ e:expr) ≫
+   --------
+   [≻ (begin-
+        (: x : τ)
+        (define x e))]]
+  [(_ x:id/type-decl e:expr) ≫
+   #:with x- (syntax-local-introduce #'x.internal-id)
+   --------
+   [≻ (ro:define x- (ann e : x.type))]]
+  [(_ x:id e:expr) ≫
+   #:with x- (generate-temporary #'x)
+   [⊢ e ≫ e- ⇒ τ]
+   #:with x-/props (transfer-props #'e- (⊢ x- : τ))
+   --------
+   [≻ (begin-
+        (define-syntax- x (make-variable-like-transformer #'x-/props))
+        (ro:define x- e-))]]
+  ;; function with no rest argument
+  [(_ (f:id [x:id : τ_in] ... [kw:keyword y:id : τ_kw e_def:expr] ...)
+      :-> τ_out
+      body ...+) ≫
+   #:with body* (syntax/loc this-syntax (begin body ...))
```

Comment:
It is on the next line, `(ann body* : τ_out)`. Because of how `ann` works with source locations, I need the `syntax/loc` on the `(begin body ...)`, not on the `(ann (begin body ...) : τ_out)`.